### PR TITLE
Tpetra: Move from 2-arg to 3-arg Kokkos::deep_copy where appropriate

### PR DESCRIPTION
**Pull request by csiefer2**
Issue: #10433
Tests: Cuda w/ and w/o UVM
Stakeholder feedback: n/a

This is a successor to #7946 which we will close when this one merges.

### CODE REVIEW

**Comment by kyungjoo-kim** [comment]
@csiefer2 @crtrott One question. One suggestion. Currently, it uses an execution space instance without any stream information. So, in such case, does Kokkos provide something like ``Kokkos::deep_copy_async(A, B)`` that is equivalent to ``Kokkos::deep_copy(exec_space(), A, B)`` ? The key point of using execution space instance is that it does not synchronize and the deep copy async is more clear for readability.

**Comment by csiefer2** [comment]
> @csiefer2 @crtrott One question. One suggestion. Currently, it uses an execution space instance without any stream information. So, in such case, does Kokkos provide something like `Kokkos::deep_copy_async(A, B)` that is equivalent to `Kokkos::deep_copy(exec_space(), A, B)` ? The key point of using execution space instance is that it does not synchronize and the deep copy async is more clear for readability.

Two points:
- Nothing currently works with streams, so all of this will need to be overhauled (again) for that anyway.
- IIRC, 3-arg deep_copy does does a cudaStreamSynchronize, so "async" here would be syntactically different.

**Comment by kyungjoo-kim** [comment]
Seeing Kokkos_CudaSpace.cpp file, providing cuda execution space instance, it does not sync. So, the syntax of using 3 arg version is equivalent to deep copy async. 
```
void DeepCopyCuda(void *dst, const void *src, size_t n) {
  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaMemcpy(dst, src, n, cudaMemcpyDefault));
}

void DeepCopyAsyncCuda(const Cuda &instance, void *dst, const void *src,
                       size_t n) {
  KOKKOS_IMPL_CUDA_SAFE_CALL(
      cudaMemcpyAsync(dst, src, n, cudaMemcpyDefault, instance.cuda_stream()));
}

void DeepCopyAsyncCuda(void *dst, const void *src, size_t n) {
  cudaStream_t s = cuda_get_deep_copy_stream();
  KOKKOS_IMPL_CUDA_SAFE_CALL(
      cudaMemcpyAsync(dst, src, n, cudaMemcpyDefault, s));
  Impl::cuda_stream_synchronize(
      s,
      Kokkos::Tools::Experimental::SpecialSynchronizationCases::
          DeepCopyResourceSynchronization,
      "Kokkos::Impl::DeepCopyAsyncCuda: Deep Copy Stream Sync");
}
```

**Comment by crtrott** [comment]
The three arg version with a cuda instance does not sync, it just enqueues the copy.

**Comment by csiefer2** [comment]
> The three arg version with a cuda instance does not sync, it just enqueues the copy.

Well, I stand corrected then.

File: packages/tpetra/core/src/Tpetra_Details_get1DConstView.hpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+#include "Tpetra_ConfigDefs.hpp"
+#include "Tpetra_Util.hpp"
+#include "Kokkos_DualView.hpp"
+#include <Teuchos_Array.hpp>
+#include <utility>
+
+namespace Tpetra {
+namespace Details {
+
+// mfh 28 Apr 2016: Sometimes we have a raw host array, and we need
```
Comment (tasmith4) [inline packages/tpetra/core/src/Tpetra_Details_get1DConstView.hpp:59]: I'm confused why a file with a comment from Mark from 2016 shows up as completely new in the diff.  Is this something we had deleted that's being re-added incorrectly?

**Review by tasmith4 — CHANGES_REQUESTED** [review]
There are a number of unqualified calls to `Kokkos::fence()` added in this PR.  I think we generally prefer `execution_space().fence()` in most instances to avoid fencing other execution spaces unnecessarily (this has bitten us before).  Are we sure that every new `Kokkos::fence()` absolutely must fence all execution spaces?

File: packages/tpetra/core/src/Tpetra_Details_get1DConstView.hpp (hunk 1 above)
Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_Details_get1DConstView.hpp:59]: Probably

**Review by cgcgcg — APPROVED** [review]

**Review by tasmith4 — APPROVED** [review]

**Comment by rppawlo** [comment]
@csiefer2 - this commit has broken panzer. It snuck through PR testing because CUDA_LAUNCH_BLOCKING was accidentally enabled in PR testing.  Might need some help fixing.

@jwillenbring @trilinos/framework - we had disabled CUDA_LAUNCH_BLOCKING in PR testing but it looks like it is now active in the vortex build. At least it is enabled in PullRequestCuda10.1.243TestingEnv.sh if that is what we are using with PR testing. Reverting this commit fixes the panzer tests.

**Comment by rppawlo** [comment]
The failing Panzer tests are:
```
The following tests FAILED:
	135 - PanzerAdaptersSTK_check_sorted_face_quads_MPI_1 (Failed)
	154 - PanzerAdaptersSTK_LocalMesh_ParallelPeriodicTests_MPI_2 (Failed)
	178 - PanzerAdaptersSTK_PoissonInterfaceTpetra_2d_MPI_4 (Failed)
	180 - PanzerAdaptersSTK_PoissonInterfaceTpetra_3d_MPI_4 (Failed)
Errors while running CTest
```

**Comment by csiefer2** [comment]
@rppawlo What configuration is failing?

**Comment by csiefer2** [comment]
Also, I would have _never_ seen this unless I was searching through my email for your name for a completely unrelated reason.

**Comment by rppawlo** [comment]
@csiefer2 = good to know. I will send direct email next time. Vortex is the issue. The cuda PR environment is setting CUDA_LAUNCH_BLOCKING=1 by accident. I think @e10harvey made a change to some of the internal PR testing scripts, but you will have to manually unset that env variable before running if you are using the scripts in Trilinos/cmake/std/sems. Those still have the variable set.

**Comment by rppawlo** [comment]
You should be able to reproduce on any cuda build though. I have reproduced on the ascicgpu machines.

**Pull request merged**